docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+    @render.text
+    def current_time():
+        reactive.invalidate_later(1)
+        return datetime.datetime.now().strftime("%H:%M:%S %p")
+
+    @ui.bind_task_button(button_id="btn")
+    @reactive.extended_task
+    async def sum_values(x, y):
+        await asyncio.sleep(5)
+        return x + y
+
+    @reactive.effect
+    @reactive.event(input.btn)
+    def btn_click():
+        sum_values(input.x(), input.y())
+
+    @render.text
+    def sum():
+        return str(sum_values.result())
+
+app = App(app_ui, server)
+```
+
+Note the `sum_values` function, which is where the actual (slow) work is done.
+It is decorated with `@reactive.extended_task`, which means that it will be run asynchronously and concurrently with other tasks, and that its result will be available as `sum_values.result()`.
+
+We've also added a `@ui.bind_task_button` decorator on top of the `@reactive.extended_task` decorator.
+This synchronizes the `ExtendedTask` object with the `ui.input_task_button` in the UI, so that the button will be in its "Processing..." state while the task is running.
+It _does not_ cause button clicks to invoke the task; we still need to do that manually, which we'll talk about next.
+
+If you use some other UI gesture or condition besides `ui.input_task_button` to invoke the task, don't include the `@ui.bind_task_button` decorator--it doesn't work with `ui.input_action_button`, for example.
+
+### Invoking an extended task
+
+Unlike a reactive effect, simply creating an extended task does not cause it to run.
+It needs to be invoked (called like a function).
+
+In this case, the `sum_values` extended task is called from the `btn_click` reactive effect (`@reactive.effect`), which runs whenever the button is clicked (`@reactive.event(input.btn)`).
+
+Notice also that the `sum_values` logic no longer reads `input.x()` and `input.y()` directly; as an extended task, it is not allowed to read reactive inputs or calcs (and attempting to do so would result in an error).
+Instead, it takes `x` and `y` as arguments, which are passed in by `btn_click`.
+
+### Using the result of an extended task
+
+The `sum` output is now a regular reactive calc, which reads `sum_values.result()` to get the result of the extended task.
+This output actually does not "wait for" the extended task to complete, exactly; instead, it will run multiple times, as the extended task goes through different states.
+For each state, `sum_values.result()` will behave differently:
+
+* **Not yet clicked:** Raises a silent exception, which will cause the `sum` output to display nothing.
+* **Running:** Raises a special type of exception that tells Shiny to keep the output in the "in progress" state.
+* **Successfully completed:** Returns the return value of `sum_values`, in this case an integer.
+* **Completed with an exception:** If `sum_values` raised an exception while processing, then re-raise that same exception, causing it to be displayed to the user in the `sum` output.
```

Comment:
```suggestion
* **Completed with an exception:** If `sum_values` raised an exception while processing, then that same exception is re-raised, causing it to be displayed to the user in the `sum` output.
```

Reply:
I'd like to keep the tense the same as the other bullets--I'll change "re-raise" in my original to "re-raises". Is that better?

Reply:
Yep, thanks